File: frontend/src/lib/text-cleaner.ts
const INVISIBLE_REGEX =
  /[\u00AD\u034F\u061C\u115F\u1160\u17B4\u17B5\u180E\u200B-\u200F\u202A-\u202E\u2060-\u2064\u2066-\u206F\u3164\uFE00-\uFE0F\uFEFF\uFFA0]/g;

/**
 * Frontend 2nd-pass cleaning: removes invisible/zero-width characters
 */
export function cleanTextLocal(text: string): string {
  return text.replace(INVISIBLE_REGEX, "");
}

/**
 * Scan text for invisible characters and return their positions
 */
export function scanTextLocal(text: string) {
  const positions: Array<{ index: number; charCode: string }> = [];
  for (let i = 0; i < text.length; i++) {
    if (INVISIBLE_REGEX.test(text[i])) {
      positions.push({
        index: i,
        charCode: `U+${text.charCodeAt(i).toString(16).toUpperCase().padStart(4, "0")}`,
      });
    }
    INVISIBLE_REGEX.lastIndex = 0;
  }
  return {
    isClean: positions.length === 0,
    totalHidden: positions.length,
    positions,
  };
}

/**
 * Mapping of invisible character codes to readable names (Korean)
 */
export const INVISIBLE_CHAR_NAMES: Record<string, string> = {
  "U+00AD": "소프트 하이픈",
  "U+034F": "조합 문자",
  "U+061C": "아랍어 마크",
  "U+200B": "영폭 공백",
  "U+200C": "영폭 비결합자",
  "U+200D": "영폭 결합자",
  "U+200E": "좌→우 마크",
  "U+200F": "우→좌 마크",
  "U+2060": "워드조이너",
  "U+FEFF": "BOM / 영폭 비분리 공백",
  "U+FE0F": "변형 선택자",
};

File: backend/app/services/text_cleaner.py
import re
import unicodedata

# GPT가 넣는 마크다운 서식 제거
MARKDOWN_PATTERN = re.compile(
    r"\*\*(.+?)\*\*"  # **bold**
    r"|__(.+?)__"      # __bold__
    r"|(?<!\w)\*(.+?)\*(?!\w)"  # *italic*
    r"|(?<!\w)_(.+?)_(?!\w)"    # _italic_
    r"|~~(.+?)~~"      # ~~strikethrough~~
    r"|^#{1,6}\s+",    # ### headings
    re.MULTILINE,
)

INVISIBLE_PATTERN = re.compile(
    "["
    "\u00ad"
    "\u034f"
    "\u061c"
    "\u115f\u1160"
    "\u17b4\u17b5"
    "\u180e"
    "\u200b-\u200f"
    "\u202a-\u202e"
    "\u2060-\u2064"
    "\u2066-\u206f"
    "\u3164"
    "\ufe00-\ufe0f"
    "\ufeff"
    "\uffa0"
    "\U000e0001"
    "\U000e0020-\U000e007f"
    "]"
)


def _strip_markdown(match: re.Match) -> str:
    """마크다운 기호는 제거하고 내용만 남김"""
    for group in match.groups():
        if group is not None:
            return group
    return ""


class TextCleaner:
    @staticmethod
    def clean(text: str) -> str:
        text = MARKDOWN_PATTERN.sub(_strip_markdown, text)
        return INVISIBLE_PATTERN.sub("", text)

    @staticmethod
    def scan(text: str) -> dict:
        positions = []
        for i, char in enumerate(text):
            if INVISIBLE_PATTERN.match(char):
                positions.append({
                    "index": i,
                    "char_code": f"U+{ord(char):04X}",
                    "char_name": unicodedata.name(char, "UNKNOWN"),
                })
        return {
            "is_clean": len(positions) == 0,
            "total_hidden": len(positions),
            "positions": positions,
            "text_length": len(text),
        }

    @staticmethod
    def clean_html(html: str) -> str:
        parts = re.split(r"(<[^>]+>)", html)
        cleaned = []
        for part in parts:
            if part.startswith("<"):
                cleaned.append(part)
            else:
                cleaned.append(INVISIBLE_PATTERN.sub("", part))
        return "".join(cleaned)
